test/AutoDiff/control_flow.swift

```diff
@@ -1,79 +1,174 @@
+    } else if x == -1337 {
+      y = 0
     } else {
+      y = x + x
+    }
+    return y
+  }
+  expectEqual(8, gradient(at: 4, in: cond2))
+  expectEqual(2, gradient(at: -10, in: cond2))
+  expectEqual(0, gradient(at: -1337, in: cond2))
+
+  func cond3(_ x: Float, _ y: Float) -> Float {
+    if x > 0 {
+      return x * y
+    }
+    return y - x
+  }
+  expectEqual((5, 4), gradient(at: 4, 5, in: cond3))
+  expectEqual((-1, 1), gradient(at: -3, -2, in: cond3))
+
+  func guard1(_ x: Float, _ y: Float) -> Float {
+    guard x > 0 else {
+      return x * x
+    }
+    return y * y
+  }
+  expectEqual((0, 10), gradient(at: 4, 5, in: guard1))
+  expectEqual((-6, 0), gradient(at: -3, -2, in: guard1))
+
+  func guard2(_ x: Float, _ y: Float) -> Float {
+    guard x > 0 else {
+      if y > 0 {
+        return x * y
+      } else if x == -1337 {
+        return x * x
+      }
+      return 0
+    }
+    return y * y
+  }
+  expectEqual((0, 10), gradient(at: 4, 5, in: guard2))
+  expectEqual((5, -1337), gradient(at: -1337, 5, in: guard2))
+  expectEqual((-2674, 0), gradient(at: -1337, -5, in: guard2))
+  expectEqual((2, -3), gradient(at: -3, 2, in: guard2))
+
+  func guard3(_ x: Float, _ y: Float) -> Float {
+    guard x > 0 else {
+      fatalError()
+    }
+    return y * y
+  }
+  expectEqual((0, 10), gradient(at: 4, 5, in: guard3))
+  expectCrash {
+    gradient(at: -3, -2, in: guard3)
+  }
+
+  func cond_empty(_ x: Float) -> Float {
+    if x > 0 {
+      // Create empty trampoline blocks.
+    }
+    return x * x
+  }
+  expectEqual(4, gradient(at: 2, in: cond_empty))
+  expectEqual(-6, gradient(at: -3, in: cond_empty))
+
+  func cond_generic<T : Differentiable & FloatingPoint>(
+    _ x: T, _ y: T
+  ) -> T {
+    if x > 0 {
+      return x
+    }
+    return y
+  }
+  expectEqual((1, 0), gradient(at: 4, 5, in: { x, y in cond_generic(x, y) }))
+  expectEqual((0, 1), gradient(at: -4, 5, in: { x, y in cond_generic(x, y) }))
+}
+
+ControlFlowTests.test("NestedConditionals") {
+  func nested1(_ x: Float) -> Float {
+    if x > 0 {
+      if x > 10 {
+        return x * x + x
+      } else {
+        return x * x
+      }
+    }
+    return x * -x
+  }
+  expectEqual(23, gradient(at: 11, in: nested1))
+  expectEqual(8, gradient(at: 4, in: nested1))
+  expectEqual(20, gradient(at: -10, in: nested1))
+
+  func nested2(_ x: Float, _ y: Float) -> Float {
+    if x > 0 {
+      if y > 10 {
+        return x * y
+      } else {
+        return x + y
+      }
+    }
+    return -y
+  }
+  expectEqual((20, 4), gradient(at: 4, 20, in: nested2))
+  expectEqual((1, 1), gradient(at: 4, 5, in: nested2))
+  expectEqual((0, -1), gradient(at: -3, -2, in: nested2))
+
+  func nested3(_ x: Float, _ y: Float) -> Float {
+    if x > 0 {
+      if y > 10 {
+        let z = x * y
+        if z > 100 {
+          return x + z
+        } else if y == 20 {
+          return z + z
+        }
+      } else {
+        return x + y
+      }
+    }
+    return -y
+  }
```

Comment:
Please add a test that manifests definitive initialization. That is, define a `let` with no initialization, and initialize it in all control flow paths. Also try `var` with some mutation in a branch.

Reply:
`func cond2` above is a definitive initialization test - I think one such test is sufficient.
I'll add `var` tests soon.

Reply:
`var` tests added in https://github.com/apple/swift/pull/25057/commits/806b4643a3c208275d4a3da8bcde6e7544fbd57b, fixed in https://github.com/apple/swift/pull/25057/commits/49937d93f87989f9d46d29a956bb42923ca143d2.